### DISCUSSION TITLE
Google indexation issues

### DISCUSSION THREAD

**Discussion by pier7529** [vercel/next.js#59155] [Help]
### Summary

During live testing, indexing issues were detected at the URL level - (5xx)

For a few weeks I have had certain pages on my site display this message in Google Search Console "During the live test, indexing problems were detected at the URL level".
I can't determine the cause of the problem (root cause) for certain pages like these (there are others) although it says "The page cannot be indexed: Server error (5xx )" but without further detail.

https://wildescapesoverland.com/fr/trip/chili
https://wildescapesoverland.com/fr/trip/japon-hokkaido

On the other hand, if I enter these urls directly into a browser I have no problem displaying them.

### Additional information

```js
The robots.txt and sitemap.xml files are easily accessible.
https://wildescapesoverland.com/fr and https://wildescapesoverland.com/en are correctly indexed.


First expert answer on https://support.google.com : 

Your server returns a 500 error, but it displays the page
You have to see on the host side what is happening, probably a small adjustment to review

there are also no index tags in your pages and JavaScript redirects between the EN and FR versions which are not optimal
```

### Example

https://wildescapesoverland.com/fr/trip/japon-hokkaido

**Comment by pier7529** [comment]
Precision : 

Display via web navigator OK 
but not for Googlebot 
curl -H "User-agent: Googlebot"  -I https://wildescapesoverland.com/fr/trip/france-alpes-gr738

**HTTP/1.1 500 Internal Server Error**
Age: 0
Cache-Control: private, no-cache, no-store, max-age=0, must-revalidate
Connection: keep-alive
Content-Length: 0
Content-Type: text/html; charset=utf-8
Date: Fri, 01 Dec 2023 20:19:52 GMT
Server: Vercel
Strict-Transport-Security: max-age=63072000
Vary: RSC, Next-Router-State-Tree, Next-Router-Prefetch, Next-Url
X-Matched-Path: /[lang]/trip/[slug]
X-Powered-By: Next.js
X-Vercel-Cache: MISS
X-Vercel-Execution-Region: iad1
X-Vercel-Id: cdg1::iad1::2xtqq-1701461991798-1b1b5146a8c2

**Comment by pier7529** [comment]
Move my internal APIs to app\api and use generateStaticParams...
Issue closed.

**Reply by jm-kyle** [reply]
I'm having the same issue, what was the solution here?

**Reply by ivandamyanov** [reply]
I have the same issue, doing a LIVE test using the search console it returns error 500, my website is totally fine. I'm hosting on Vercel and using Cloudflare DNS

**Reply by Alban-i** [reply]
I had the same issue, and it was related to localization in middleware, which was not returning any locale when Google was working on indexing.

So I applied this solution:
```typescript
import { NextResponse } from 'next/server';
import type { NextRequest } from 'next/server';

import { i18n } from '@/i18n.config';

import { match as matchLocale } from '@formatjs/intl-localematcher';
import Negotiator from 'negotiator';
import { updateSession } from './providers/supabase/middleware';

function getLocale(request: NextRequest): string {
  try {
    // Convert request headers to a negotiable format
    const negotiatorHeaders: Record<string, string> = {};
    request.headers.forEach((value, key) => (negotiatorHeaders[key] = value));

    // Ensure locales are accessible
    const locales = [...i18n.locales]; // Clone to avoid mutating readonly array
    const languages = new Negotiator({
      headers: negotiatorHeaders,
    }).languages();

    // Match the locale
    const locale = matchLocale(languages, locales, i18n.defaultLocale);

    // Ensure a valid locale is always returned
    return locale || i18n.defaultLocale;
  } catch (error) {
    console.error('Error determining locale:', error);

    // Fallback to default locale in case of error
    return i18n.defaultLocale;
  }
}

//  ############################ MIDDLEWARE ###################################
export async function middleware(request: NextRequest) {
  const pathname = request.nextUrl.pathname;
  const pathnameIsMissingLocale = i18n.locales.every(
    (locale) => !pathname.startsWith(`/${locale}/`) && pathname !== `/${locale}`
  );

  // Redirect if there is no locale
  if (pathnameIsMissingLocale) {
    const locale = getLocale(request);
    return NextResponse.redirect(
      new URL(
        `/${locale}${pathname.startsWith('/') ? '' : '/'}${pathname}`,
        request.url
      )
    );
  }

  // Supabase SSR
  return await updateSession(request);
}

export const config = {
  matcher: [
    '/((?!api|_next/static|_next/image|robots.txt|sitemap.xml|favicon.ico|auth|public|manifest.json|manifest|apple-icon.png|images).*)',
  ],
};
```
